### ISSUE TITLE
HIDPI per monitor

### ISSUE THREAD

**Issue by MartinX3** [mate-desktop/mate-desktop#323]
#### Expected behaviour
HIDPI enabled or disabled per monitor.

#### Actual behaviour
HIDPI enabled or disabled globally.

#### Steps to reproduce the behaviour
![grafik](https://user-images.githubusercontent.com/4518092/41498311-667ca700-716a-11e8-9bd3-668ef716010b.png)

#### MATE general version
1.20

#### Package version
1.20.1-2ubuntu1

#### Linux Distribution
Linux Mint 19 x64 Beta

**Comment by MartinX3** [comment]
Thank you very much!
But I think, I lower the resolution of the desktop and deactivate HIDPI.
I want to do my work with linux, not working for linux. :P

**Comment by EmilyBjoerk** [comment]
@raveit65 the xrandr solution is a bad one for the following reasons:

* 8x the computational effort for the scaled display (render at double resolution (4x) then rescale down another +4x)
* The scaling will mess up subpixel rendering on fonts that now look awfull
* The scaling will also make everything slightly blurry and out of place due to the bi-linear scaling.

**Comment by raveit65** [comment]
> * 8x the computational effort for the scaled display (render at double resolution (4x) then rescale down another +4x)
> * The scaling will mess up subpixel rendering on fonts that now look awfull
> * The scaling will also make everything slightly blurry and out of place due to the bi-linear scaling.
> * Only integer scaling values supported.

You're right, i dream about to have 2 - 10 monitors with same HIDPI Resolutions :)
- Than i can support and debug every kind of consumers wishes.
- than i can build a video-wall in my room, isn't that nice, isn't it?
- i am not said about not able to fix your problems :/

**Comment by EmilyBjoerk** [comment]
Well that was passive aggressive and borderline childish.

HiDPI monitors are not that expensive any more and are becoming increasingly common which is evident by these HiDPI questions popping up more and more. Non-support for mixed DPI and poor support for single HiDPI is making it harder to use MATE which is otherwise a nice DE.

**Comment by vkareh** [comment]
I didn't see the comment as childish, but I'm not a native speaker, so I don't know.

I would also love a wall of HiDPI monitors. But not everyone can afford even a single one.

My work laptop is not HiDPI, but my monitor is. To use both I have to rely on xrandr. Yes, it's slower, but it's the only solution at the moment until GTK v4. Unless we do re-scaling whenever you move a window to another monitor, but that will very likely be worse performance than xrandr :/

**Comment by lukefromdc** [comment]
Porting to GTK4 would be a major problem. It won't be stable for a while, which would put our small team right back on the treadmill GTK3 used to be. A lot of functions we use that were deprecated (e.g GtkStatusIcon and GtkPlug/Socket) have no direct replacements, and worst of all most custom-drawn widgets including the desktop have to be refactored. That last consideration is why GNOME branched GTK4 (originally as GTK 3.80something) when the GSK (scene graph) commits were merged, and these were kept all the way out of GTK3. Firefox and Libreoffice had just ported to GTK3, using almost entirely custom widgets. GSK would have meant a total do-over for both of them because of the custom widgets and drawing.

We would have a lot of rework in mate-panel and caja for GTK4

**Comment by raveit65** [comment]
> Well that was passive aggressive and borderline childish.

Closing this as getting personal!

**Comment by MartinX3** [comment]
Please don't close the issue.
Not everyone wants to Rebuy every monitor for the sake of having only HiDPI monitors (money and environmental pollution causes).

**Comment by raveit65** [comment]
@MartinX3 
Sorry , closing was a mistake , But issue is already open.

**Comment by lukefromdc** [comment]
Well, I just got per-monitor scaling functioning by a script on my box, using this code w comments for adapting it:
```

xrandr --output VGA-0 --auto --scale 2x2 --right-of HDMI-A-0

# https://wiki.archlinux.org/index.php/HiDPI#Multiple_displays
#use xrandr -q to identify displays
```
Since this is xrandr it should be possible to add support for the --scale to the mate-settings-daemon xrandr applet, and we would get something not a lot of DE's support in X11 at the moment

**Comment by lukefromdc** [comment]
On further testing I found that on my very powerful desktop, using xrandr to set the secondary monitor to a non hidpi resolution alongside a 4K monitor caused severe performance issues, at least with a small first-gen AMD GCN card forced onto the AMDGPU driver. There were issues with cursor blinking in and out on QT apps, whether with the hardware or the software cursor options. The main offender though was laggy video playback on the secondary monitor, which is simply not acceptable when video editing is the intended use.

Thus, there would be limited value in using xrandr to support dissimilar monitor scaling in mate-settings-daemon. Theoretically it might be possible instead to rewrite every app to detect resolution of the monitor it is on and set the scaling(calculating it once-only)  that way on a per-app basis, but this would be hugely complex, beyond the limits of our small team, and would do nothing for non-MATE apps. 

Since X itself does not understand scaling and GTK is simply applying a multiplier, the right place for a fix is in GTK itself, but since that's a new feature it might be limited to GTK4, no idea if it would ever get backported.

**Comment by EmilyBjoerk** [comment]
DPI and scaling are easily confused concepts that unfortunately have been historically conflated.  DPI is used for making sure that elements are consistently scaled between displays. E.g a 1cm line renders as 1cm long regardless of monitor. And scaling is a shortcut for getting everything to be bigger e.g. when you are using a TV as your monitor and watching it from a distance. You want things to look bigger, this is scaling. You can use DPI settings to change the size of UI elements to make them bigger, but you shouldn't, as this messes up a bunch of other things such as scale correct printing previews etc. The DPI is an intrinsic attribute of the monitor that should be respected, hence why it's so horrid that X11 forces all monitors to 96 DPI and depends on userland tools such as xrandr and WMs to compute the correct DPI and set it. They did this because otherwise TVs and projectors would render poorly when the real solution is for TV and projectors to auto detect and set a sensible scaling value instead of messing up everyone else's displays, and also because some webpages in some old versions of firefox (or was it chromium?) would render "wrong" (crap in crap out should apply imho)... </rant>

Unfortunately both Linux (well X and GTK anyway) and Windows got this wrong historically and we seem to be stuck with historical mistakes.

I believe it is possible to hack GTK so that you can set the DPI (not scaling) per application (they already use DPI from xrandr, just need a mechanism for overriding it and passing the value) and I also believe that it would be possible for WMs to set this DPI per GTK application in some way. I'm not so familiar with how WMs communicated with their client applications, probably through some X/WM protocol that could be extended?

**Comment by lukefromdc** [comment]
Also keep in mind, the application might have windows open on more than one monitor, so per-application helps in many but not all cases. Still, so long as we can avoid repeat calculations in xrandr with that huge performance penatly (as per my tests), this might help. Caja (which renders the desktop) will be one application that won't benefit from per-app scaling, as caja navigation windows on a secondary monitor will still be bound to the same scale as the desktop icons on the primary monitor. We need a way to detect which monitor we are on.
